sonar-batch/src/main/java/org/sonar/batch/report/TestExecutionAndCoveragePublisher.java

```diff
@@ -0,0 +1,128 @@
+import org.sonar.api.batch.fs.internal.DefaultInputFile;
+import org.sonar.api.test.CoverageBlock;
+import org.sonar.api.test.MutableTestCase;
+import org.sonar.api.test.MutableTestPlan;
+import org.sonar.api.test.TestCase;
+import org.sonar.batch.index.BatchResource;
+import org.sonar.batch.index.ResourceCache;
+import org.sonar.batch.protocol.Constants.TestStatus;
+import org.sonar.batch.protocol.output.BatchReport;
+import org.sonar.batch.protocol.output.BatchReport.CoverageDetail;
+import org.sonar.batch.protocol.output.BatchReport.Test;
+import org.sonar.batch.protocol.output.BatchReportWriter;
+import org.sonar.core.test.TestPlanBuilder;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class TestExecutionAndCoveragePublisher implements ReportPublisherStep {
+
+  private final ResourceCache resourceCache;
+  private final TestPlanBuilder testPlanBuilder;
+
+  public TestExecutionAndCoveragePublisher(ResourceCache resourceCache, TestPlanBuilder testPlanBuilder) {
+    this.resourceCache = resourceCache;
+    this.testPlanBuilder = testPlanBuilder;
+  }
+
+  @Override
+  public void publish(BatchReportWriter writer) {
+    for (final BatchResource resource : resourceCache.all()) {
+      if (!resource.isFile()) {
+        continue;
+      }
+
+      DefaultInputFile inputFile = (DefaultInputFile) resource.inputPath();
+      if (inputFile.type() != Type.TEST) {
+        continue;
+      }
+
+      final MutableTestPlan testPlan = testPlanBuilder.get(MutableTestPlan.class, inputFile.key());
+      if (testPlan == null || Iterables.isEmpty(testPlan.testCases())) {
+        continue;
+      }
+
+      final Set<String> testNamesWithCoverage = new HashSet<>();
+
+      writer.writeTests(resource.batchId(), Iterables.transform(testPlan.testCases(), new Function<MutableTestCase, BatchReport.Test>() {
+
+        private BatchReport.Test.Builder builder = BatchReport.Test.newBuilder();
+
+        @Override
+        public Test apply(MutableTestCase testCase) {
+          builder.clear();
+          builder.setName(testCase.name());
+          if (testCase.doesCover()) {
+            testNamesWithCoverage.add(testCase.name());
+          }
+          Long durationInMs = testCase.durationInMs();
+          if (durationInMs != null) {
+            builder.setDurationInMs(durationInMs);
+          }
+          String msg = testCase.message();
+          if (msg != null) {
+            builder.setMsg(msg);
+          }
+          String stack = testCase.stackTrace();
+          if (stack != null) {
+            builder.setStacktrace(stack);
+          }
+          TestCase.Status status = testCase.status();
+          if (status != null) {
+            builder.setStatus(TestStatus.valueOf(status.name()));
+          }
+          return builder.build();
+        }
+
+      }));
+
+      writer.writeCoverageDetails(resource.batchId(), Iterables.transform(testNamesWithCoverage, new Function<String, CoverageDetail>() {
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Reduce this anonymous class number of lines from 23 to at most 20, or make it a named class.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1188)